feat: liquid alpha

distributedstatemachine opened the pull request:
## Description
<!--
  Please provide a brief description of the changes introduced by this pull request.
-->

This PR implements 

- Liquid Alpha feature. 
- moves the maths unit tests to the tests folder 

Supercedes: #503 




## Related Issue(s)

- Closes #417 

## Type of Change
<!--
Please check the relevant options:
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

## Breaking Change

If this PR introduces a breaking change, please provide a detailed description of the impact and the migration path for existing applications.

## Checklist

<!--
Please ensure the following tasks are completed before requesting a review:
-->

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run `cargo fmt` and `cargo clippy` to ensure my code is formatted and linted correctly
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots (if applicable)

Please include any relevant screenshots or GIFs that demonstrate the changes made.

## Additional Notes

Please provide any additional information or context that may be helpful for reviewers.

Commit:
feat: liquid alpha

Commit:
feat: clippy , zepter

Commit:
chore: bump spec

JohnReedV committed:
add import

Commit:
feat: use u16:MAX norm , update tests

Commit:
fix tests lints

Commit:
fix: check migrations for right environments

Commit:
fix: feature propagation

Commit:
chore: saturating div , update ema dense

Commit:
lint

sam0x17 commented on pallets/subtensor/tests/math.rs:
```diff
@@ -0,0 +1,2344 @@
+    let result: Vec<I32F32> = vec![fixed(1.) / fixed(2.), fixed(1.) / fixed(3.), fixed(5.)];
+    assert_eq!(result, vecdiv(&x, &y));
+}
+
```
maybe also check that `vecdiv`'s behavior with division by zero is correct, like:

```suggestion

#[test]
fn test_math_vecdiv_zero_division() {
    let x: Vec<I32F32> = vec_to_fixed(&[1., 2., 3.]);
    let y: Vec<I32F32> = vec_to_fixed(&[0., 0., 0.]);
    let result = vecdiv(&x, &y);
    assert!(result.iter().all(|&val| val.is_nan())); // Expect NaN or error
}

```

distributedstatemachine replied:
> maybe also check that `vecdiv`'s behavior with division by zero is correct, like:

This isnt code i added, so I wouldnt want to modify this test

sam0x17 commented on pallets/subtensor/tests/math.rs:
```diff
@@ -0,0 +1,2344 @@
+    ];
+    assert_mat_compare(&mat, &vec_to_mat_fixed(&target, 4, false), epsilon);
+}
+
```
Perhaps a check that `normalize` handles cases where the input vector sum is zero to avoid division by zero?

```suggestion

#[test]
fn test_math_normalize_zero_sum() {
    let epsilon: I32F32 = I32F32::from_num(0.0001);
    let x: Vec<I32F32> = vec![I32F32::from_num(0.0), I32F32::from_num(0.0)];
    let y: Vec<I32F32> = normalize(&x);
    assert_vec_compare(&x, &y, epsilon); // Expect y to handle zero-sum case gracefully
}


```

distributedstatemachine replied:
> Perhaps a check that `normalize` handles cases where the input vector sum is zero to avoid division by zero?

also an existing test, that was not added in this pr

sam0x17 commented on pallets/subtensor/tests/math.rs:
```diff
@@ -0,0 +1,2344 @@
+    );
+}
+
+#[test]
+fn test_math_fixed_to_u16() {
+    let expected = u16::MIN;
+    assert_eq!(fixed_to_u16(I32F32::from_num(expected)), expected);
+
+    let expected = u16::MAX / 2;
+    assert_eq!(fixed_to_u16(I32F32::from_num(expected)), expected);
+
+    let expected = u16::MAX;
+    assert_eq!(fixed_to_u16(I32F32::from_num(expected)), expected);
+}
```
maybe test handling when we are over the `u16` capacity? does it saturate?

distributedstatemachine replied:
noooooooot my testtt

sam0x17 commented on pallets/subtensor/tests/math.rs:
```diff
@@ -0,0 +1,2344 @@
+        I32F32::from_num(0),
+    );
+}
+
```
Also would be nice to check that `normalize` is handling single and empty inputs correctly:

```suggestion

#[test]
fn test_empty_and_single_element_cases() {
    let empty_vec: Vec<I32F32> = vec![];
    assert_eq!(normalize(&empty_vec), empty_vec);

    let single_element_vec: Vec<I32F32> = vec![I32F32::from_num(1.0)];
    assert_eq!(normalize(&single_element_vec), single_element_vec);
}

```

sam0x17 reviewed (dismissed):
looks good, a few questions and potential tests that can be added that I have provided as commit suggestions for your convenience

JohnReedV committed:
unite alpha getters/setters & add extrinsic

Combines the getters and setters for alpha low/high

Removes alpha high/low hyperparams

adds an extrinsic to set alpha high/low

adds alpha values to subnetinfo

distributedstatemachine commented on pallets/subtensor/src/epoch.rs:
```diff
@@ -1185,4 +1181,30 @@ impl<T: Config> Pallet<T> {
         }
     }
+
+    pub fn do_set_alpha_values(origin: T::RuntimeOrigin, netuid: u16, alpha_low: u16, alpha_high: u16) -> Result<(), DispatchError> {
```
this seems to be missing the admin extrinsic to set it

distributedstatemachine commented on pallets/subtensor/src/epoch.rs:
```diff
@@ -1185,4 +1181,30 @@ impl<T: Config> Pallet<T> {
+
+    pub fn do_set_alpha_values(origin: T::RuntimeOrigin, netuid: u16, alpha_low: u16, alpha_high: u16) -> Result<(), DispatchError> {
+        // --- 1. Ensure the function caller is a signed user.
+        let _coldkey = ensure_signed(origin)?;
```
do we let the variable binding ? 

```
ensure_signed(origin)
```

JohnReedV committed:
ensure caller is subnet owner & tests

JohnReedV committed:
move set_alpha extrinsic to admin-utils

JohnReedV committed:
separate params in SubnetInfo

Commit:
test: compute alpha for 256 miners

Commit:
lints , use sat maths for 256 miner alpha test

Mananitade commented on pallets/admin-utils/tests/tests.rs:
```diff
@@ -1225,39 +1228,37 @@ fn test_sudo_get_set_alpha() {
+            alpha_low,
+            alpha_high
+        ));
+        let (grabbed_alpha_low, grabbed_alpha_high): (u16, u16) =
```
This is the line that I'm referring to @distributedstatemachine

Mananitade replied:
In test_clamp_alpha_values(), we call the "get_alpha_value" function to get the "grabbed_alpha_low", there is such analogous call in for "alpha" test_compute_alpha_values_256_miners()

distributedstatemachine replied:
As explained in discord , this is a different test and isnt testing the calculation of alpha values , but setting them with the `sudo_set_alpha_values` admin extrinsic

distributedstatemachine replied:
https://github.com/opentensor/subtensor/pull/552/files/2e51ebdf446ce58fd2ca81e72329354e2955bcdb..b8364014f36b6a72357881cacb0f3462a3fde5a0#diff-a89ee0066188e9aac5a009116c3c587c2fe6d6b4eff0dfd66bf50c00e0c82350R1223

Mananitade replied:
Got it, it would be great to add a note in comments to let the reader know the scope of the test, i.e., "setting values" versus testing the calculation of alpha values.

JohnReedV committed:
fix subnet info

open-junius commented on pallets/admin-utils/src/lib.rs:
```diff
@@ -996,6 +996,45 @@ pub mod pallet {
+        /// Sets values for liquid alpha
+        #[pallet::call_index(51)]
+        #[pallet::weight((0, DispatchClass::Operational, Pays::No))]
+        pub fn sudo_set_alpha_values(
```
It is better to split into two extrinsics, which will make the bittensor easier to implement the command with unified format.

distributedstatemachine replied:
@open-junius from the UX point ,we decided to make this 1 call.

distributedstatemachine committed:
Merge branch 'main' into feat/liquid_alpha_2

Commit:
chore: merge conflicts , safe maths

Commit:
chore: bump spec
